services/replayer/blocks_provider.go

```diff
@@ -0,0 +1,112 @@
+package replayer
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/storage"
+	evmTypes "github.com/onflow/flow-go/fvm/evm/types"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+)
+
+type blockSnapshot struct {
+	*BlocksProvider
+	block models.Block
+}
+
+var _ evmTypes.BlockSnapshot = (*blockSnapshot)(nil)
+
+func (bs *blockSnapshot) BlockContext() (evmTypes.BlockContext, error) {
+	return evmTypes.BlockContext{
+		ChainID:                evmTypes.EVMChainIDFromFlowChainID(bs.chainID),
+		BlockNumber:            bs.block.Height,
+		BlockTimestamp:         bs.block.Timestamp,
+		DirectCallBaseGasUsage: evmTypes.DefaultDirectCallBaseGasUsage,
+		DirectCallGasPrice:     evmTypes.DefaultDirectCallGasPrice,
+		GasFeeCollector:        evmTypes.CoinbaseAddress,
+		GetHashFunc: func(n uint64) gethCommon.Hash {
+			// For block heights greater than or equal to the current,
+			// return an empty block hash.
+			if n >= bs.block.Height {
+				return gethCommon.Hash{}
+			}
+			// If the given block height, is more than 256 blocks
+			// in the past, return an empty block hash.
+			if bs.block.Height-n > 256 {
+				return gethCommon.Hash{}
+			}
```

Comment:
```suggestion
```

I think `bs.blocks.GetByHeight(n)` handles this case already

Reply:
If I remove this line, then the relevant test case fails with:
```bash
Not equal: 
expected: 0x0000000000000000000000000000000000000000000000000000000000000000
actual  : 0x5bc86e647028e549fd4e95bd845fee4ea79f4e32eef71ebe2d0aa5cf3dc64985
```

That's because `bs.blocks.GetByHeight(n)` is using the `Blocks` DB of the EVM Gateway, which contains all blocks since genesis block. See: https://github.com/onflow/flow-evm-gateway/blob/main/storage/pebble/blocks.go#L124-L144.
We keep track of all indexed blocks in EVM Gateway, so we can return the proper response for endpoints such as:
- `eth_getBlockByNumber`
- `eth_getBlockByHash`
- `eth_getBlockReceipts` etc